src/plugins/data/public/search/session/session_service.ts

```diff
@@ -0,0 +1,247 @@
+    restoreState: UrlGeneratorStateMapping[ID]['State'];
+  }>;
+}
+
+/**
+ * Responsible for tracking a current search session. Supports only a single session at a time.
+ */
+export class SessionService implements ISessionService {
+  public readonly state$: Observable<SessionState>;
+  private readonly state: SessionStateContainer<TrackSearchDescriptor>;
+
+  private searchSessionRestorationInfoProvider?: SearchSessionRestorationInfoProvider;
+  private appChangeSubscription$?: Subscription;
+  private curApp?: string;
+
+  constructor(
+    initializerContext: PluginInitializerContext<ConfigSchema>,
+    getStartServices: StartServicesAccessor,
+    private readonly sessionsClient: ISessionsClient,
+    { freezeState = true }: { freezeState: boolean } = { freezeState: true }
+  ) {
+    const { stateContainer, sessionState$ } = createSessionStateContainer<TrackSearchDescriptor>({
+      freeze: freezeState,
+    });
+    this.state$ = sessionState$;
+    this.state = stateContainer;
+
+    getStartServices().then(([coreStart]) => {
+      // Apps required to clean up their sessions before unmounting
+      // Make sure that apps don't leave sessions open.
+      this.appChangeSubscription$ = coreStart.application.currentAppId$.subscribe((appName) => {
+        if (this.state.get().sessionId) {
+          const message = `Application '${this.curApp}' had an open session while navigating`;
+          if (initializerContext.env.mode.dev) {
+            // TODO: This setTimeout is necessary due to a race condition while navigating.
+            setTimeout(() => {
+              coreStart.fatalErrors.add(message);
+            }, 100);
+          } else {
+            // eslint-disable-next-line no-console
+            console.warn(message);
+            this.clear();
+          }
+        }
+        this.curApp = appName;
+        this.setSearchSessionRestorationInfoProvider(undefined);
+      });
+    });
+  }
+
+  /**
+   * Set a provider of info about current session
+   * This will be used for creating a background session saved object
+   * @param searchSessionRestorationInfoProvider
+   */
+  public setSearchSessionRestorationInfoProvider<ID extends UrlGeneratorId = UrlGeneratorId>(
+    searchSessionRestorationInfoProvider: SearchSessionRestorationInfoProvider<ID> | undefined
+  ) {
+    this.searchSessionRestorationInfoProvider = searchSessionRestorationInfoProvider;
+  }
+
+  /**
+   * Used to track pending searches within current session
+   *
+   * @param sessionId - sessionId that search belongs to. If not matching current session id, then search is not tracked
+   * @param searchDescriptor - uniq object that will be used to untrack the search
+   * @returns untrack function
+   */
+  public trackSearch(
+    sessionId: string | undefined,
+    searchDescriptor: TrackSearchDescriptor
+  ): () => void {
+    if (!sessionId) return () => {};
+    if (sessionId !== this.getSessionId()) return () => {};
```

Comment:
I'm wondering if we should throw in this case rather than silently ignore it. If a `sessionId` is passed in, and it doesn't match the current `sessionId`, isn't this an error case?

Reply:
I agree that this looks a bit weird. Initially wanted to make consumer code simpler. 
I changed as suggested, so now consumer has to check if they want to track search.